obstore/python/obstore/fsspec.py

```diff
@@ -0,0 +1,159 @@
+            *args, asynchronous=asynchronous, loop=loop, batch_size=batch_size, **kwargs
+        )
+
+    async def _rm_file(self, path, **kwargs):
+        return await obs.delete_async(self.store, path)
+
+    async def _cp_file(self, path1, path2, **kwargs):
+        return await obs.copy_async(self.store, path1, path2)
+
+    async def _pipe_file(self, path, value, **kwargs):
+        return await obs.put_async(self.store, path, value)
+
+    async def _cat_file(self, path, start=None, end=None, **kwargs):
+        if start is None and end is None:
+            resp = await obs.get_async(self.store, path)
+            return await resp.bytes_async()
+
+        if start is not None and end is not None:
+            return await obs.get_range_async(
+                self.store, path, offset=start, length=end - start
+            )
+
+        raise NotImplementedError("todo: handle open-ended ranges")
+
+    async def _cat_ranges(
+        self,
+        paths: List[str],
+        starts: List[int],
+        ends: List[int],
+        max_gap=None,
+        batch_size=None,
+        on_error="return",
+        **kwargs,
+    ):
+        # TODO: need to go through this again and test it
+        per_file_requests: Dict[str, List[Tuple[int, int, int]]] = defaultdict(list)
+        for idx, (path, start, end) in enumerate(zip(paths, starts, ends)):
+            per_file_requests[path].append((start, end, idx))
+
+        futs: List[Coroutine[Any, Any, List[bytes]]] = []
+        for path, ranges in per_file_requests.items():
+            offsets = [r[0] for r in ranges]
+            lengths = [r[1] - r[0] for r in ranges]
+            fut = obs.get_ranges_async(
+                self.store, path, offsets=offsets, lengths=lengths
+            )
+            futs.append(fut)
+
+        result = await asyncio.gather(*futs)
+
+        output_buffers: List[bytes] = [b""] * len(paths)
+        for per_file_request, buffers in zip(per_file_requests.items(), result):
+            path, ranges = per_file_request
+            for buffer, ranges_ in zip(buffers, ranges):
+                initial_index = ranges_[2]
+                output_buffers[initial_index] = buffer
+
+        return output_buffers
+
+    async def _put_file(self, lpath, rpath, **kwargs):
+        with open(lpath, "rb") as f:
+            await obs.put_async(self.store, rpath, f)
+
+    async def _get_file(self, rpath, lpath, **kwargs):
+        with open(lpath, "wb") as f:
+            resp = await obs.get_async(self.store, rpath)
+            async for buffer in resp.stream():
+                f.write(buffer)
+
+    async def _info(self, path, **kwargs):
+        head = await obs.head_async(self.store, path)
+        return {
+            # Required of `info`: (?)
+            "name": head["path"],
+            "size": head["size"],
+            "type": "directory" if head["path"].endswith("/") else "file",
+            # Implementation-specific keys
+            "e_tag": head["e_tag"],
```

Comment:
We have two different capitalizations of `e_tag` and `ETag`. Is one better than the other?

Reply:
It depends on how closely you want to match hat fsspec gives you. It had ETag, which is what AWS calls it in the response data. I don't mind...

Reply:
This integration should match fsspec semantics as closely as possible. The rest of obstore will probably stick with returning `e_tag` because that's what `object_store` uses under the hood. But if fsspec users would expect `ETag`, then we should change it here.

Reply:
> if fsspec users would expect `ETag`, then we should change it here.

It looks like [s3fs uses `ETag`](https://github.com/fsspec/s3fs/blob/f3f63cbfbfe71a4355abd63cafd8c678c4a5a0af/s3fs/tests/test_s3fs.py#L2468), but [adlfs uses `etag`](https://github.com/fsspec/adlfs/blob/7f06dbdd410224047df09d0ef9e5e9913c64bf7d/adlfs/spec.py#L61). I would've hoped for more consistency here. Ideally this should be defined in the fsspec spec.

Reply:
We could take this chance to make a guarantee about available fields... fsspec does have the `ukey` method on a filesystem that can be the etag output, but it would be best to have some guarantees about field names and types from here on. That would probably mean making dict-like subclasses for each (most) implementations, which derives something like etag and created time from whatever the remote gives.

Reply:
(GCS provides "etag", as well as md5 and crc)

Reply:
In the absence of a standard from fsspec, I'd opt to use `e_tag` to be consistent with the rest of the library, and we can make a breaking change later. `ETag` is still used here: https://github.com/developmentseed/obstore/pull/63/files/0a0a2fc7d7119c290b33e1b1985f625dfbb2b79e#diff-36786859c2825a3983ff9208d66ac33cc191b38bf6fc1e562a56a2b073ae229cR135